### PR TITLE
[SPARK-14343][SQL] Proper column pruning for text data source

### PR DESCRIPTION
**Pull request by liancheng**
## What changes were proposed in this pull request?

Text data source ignores requested schema, and may give wrong result when the only data column is not requested. This may happen when only partitioning column(s) are requested for a partitioned text table.
## How was this patch tested?

New test case added in `TextSuite`.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/text/TextFileFormat.scala [hunk 1]
```diff
@@ -96,16 +96,23 @@ class TextFileFormat extends FileFormat with DataSourceRegister {
...
-        unsafeRowWriter.write(0, line.getBytes, 0, line.getLength)
-        unsafeRow.setTotalSize(bufferHolder.totalSize())
-        unsafeRow
+      val reader = new HadoopFileLinesReader(file, broadcastedHadoopConf.value.value)
+
+      if (requiredSchema.isEmpty) {
+        val emptyUnsafeRow = new UnsafeRow(0)
+        reader.map(_ => emptyUnsafeRow)
+      } else {
+        val unsafeRow = new UnsafeRow(1)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/text/TextFileFormat.scala]: should we add an assert to guarantee `requiredSchema` has less than one element?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/text/TextFileFormat.scala [hunk 2]
```diff
@@ -96,16 +96,23 @@ class TextFileFormat extends FileFormat with DataSourceRegister {
...
-
-      new HadoopFileLinesReader(file, broadcastedHadoopConf.value.value).map { line =>
-        // Writes to an UnsafeRow directly
-        bufferHolder.reset()
-        unsafeRowWriter.write(0, line.getBytes, 0, line.getLength)
-        unsafeRow.setTotalSize(bufferHolder.totalSize())
-        unsafeRow
+      val reader = new HadoopFileLinesReader(file, broadcastedHadoopConf.value.value)
+
+      if (requiredSchema.isEmpty) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/text/TextFileFormat.scala]: should we put this `if` out of the returned function?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/text/TextFileFormat.scala (hunk 1 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/text/TextFileFormat.scala]: Added at the beginning of this method.

**Comment by cloud-fan** [comment]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/text/TextFileFormat.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/text/TextFileFormat.scala]: nvm, we can't save the reader creation anyway

**Comment by liancheng** [comment]
Thanks for the review! Merging to master and branch-2.0.

**Pull request closed without merging**